File: admin/src/main/java/com/project/admin/controller/TechBlogController.java
package com.project.admin.controller;

import com.project.admin.controller.dto.request.techblog.TechBlogListRequest;
import com.project.admin.controller.dto.request.techblog.TechBlogUpdateStatusRequest;
import com.project.admin.controller.dto.response.PageResponse;
import com.project.admin.controller.dto.response.techblog.TechBlogDetailResponse;
import com.project.admin.controller.dto.response.techblog.TechBlogListResponse;
import com.project.admin.controller.spec.TechBlogApi;
import com.project.admin.service.TechBlogService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RequiredArgsConstructor
@RestController
@RequestMapping("/techblog")
public class TechBlogController implements TechBlogApi {

    private final TechBlogService techBlogService;

    @GetMapping
    public ResponseEntity<PageResponse<TechBlogListResponse>> getTechBlogs(
            @Valid TechBlogListRequest request
    ) {
        Page<TechBlogListResponse> results = techBlogService.getTechBlogs(request)
                .map(TechBlogListResponse::of);

        PageResponse<TechBlogListResponse> result = PageResponse.of(results, request.page());
        return ResponseEntity.ok(result);
    }

    @GetMapping("/{id}")
    public ResponseEntity<TechBlogDetailResponse> getTechBlog(@PathVariable String id) {
        TechBlogDetailResponse response = TechBlogDetailResponse.of(
                techBlogService.getTechBlog(Long.valueOf(id))
        );

        return ResponseEntity.ok(response);
    }

    @PatchMapping("/status")
    public ResponseEntity<Void> updateStatus(@RequestBody TechBlogUpdateStatusRequest request) {

        List<Long> ids = request.ids().stream()
                .map(Long::valueOf)
                .toList();

        techBlogService.updateStatus(ids, request.status());
        return ResponseEntity.ok().build();
    }

}

File: api/src/main/kotlin/com/project/api/controller/TechBlogController.kt
package com.project.api.controller

import com.project.api.constants.Region
import com.project.api.controller.dto.CompanyListResponse
import com.project.api.controller.dto.PopularCompanyListResponse
import com.project.api.controller.dto.TechBlogDetailResponse
import com.project.api.controller.dto.TechBlogListResponse
import com.project.api.controller.dto.TechBlogSearchCondition
import com.project.api.controller.dto.TechBlogSourceSearchCondition
import com.project.api.controller.spec.TechBlogApiSpecification
import com.project.api.service.TechBlogService
import jakarta.validation.Valid
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/tech-blogs")
class TechBlogController(
    private val techBlogService: TechBlogService,
) : TechBlogApiSpecification {

    @GetMapping("/companies")
    override fun getCompanies(
        @RequestParam(required = false) region: Region?,
    ): CompanyListResponse {
        val results = techBlogService.getCompanies(region)
        return CompanyListResponse.from(results = results)
    }
    @GetMapping("/{id}")
    override fun getTechBlog(@PathVariable("id") id: String): TechBlogDetailResponse {

      val longId = id.toLong()
      return  TechBlogDetailResponse.of(techBlogService.getTechBlog(longId));
    }

    @GetMapping("/companies/popular")
    override fun getPopularCompanies(): PopularCompanyListResponse {
        val results = techBlogService.getPopularCompanies()
        return PopularCompanyListResponse.from(results = results)
    }

    @GetMapping
    override fun getTechBlogs(
        @Valid searchCondition: TechBlogSearchCondition,
    ): TechBlogListResponse {
        val results = techBlogService.getTechBlogs(searchCondition.keyword, searchCondition.page, searchCondition.size)
        return TechBlogListResponse.from(results = results)
    }

    @GetMapping("/sources")
    override fun getTechBlogsBySource(
        @Valid searchCondition: TechBlogSourceSearchCondition,
    ): TechBlogListResponse {
        val results = techBlogService.getTechBlogsBySource(searchCondition.source, searchCondition.page, searchCondition.size)
        return TechBlogListResponse.from(results = results)
    }
}
